src/main/java/org/jvnet/solaris/libzfs/LibZFS.java

```diff
@@ -47,9 +52,108 @@
 public class LibZFS implements ZFSContainer {
 
     private libzfs_handle_t handle;
+/*
+ * Track features available in current host ZFS ABI so we can use specific JNA
+ * signatures. These will be populated as string tuples (key=value) as defined
+ * in this library privately; beside auto-guesswork they can be populated from
+ * environment variables (via appserver/init-script setup for hosting Jenkins)
+ * as a fallback - but being private, admins should expect to reconfigure such
+ * strings when this library gets updated... in reality, auto-guesswork should
+ * (be implemented at all and then) do a good job.
+ * Note: in this context, the "legacy" ABI refers to function signatures that
+ * were in the OpenSolaris codebase prior to split between Oracle and illumos
+ * and subsequently OpenZFS. The "openzfs" refers to the state of OpenZFS code
+ * "Git HEAD" at the time of relevant edit of these libzfs.jar sources. Keep
+ * in mind that downstream projects may lag accepting such changes, maybe for
+ * years. For this reason, we keep a toggle for every function where we care
+ * about signature differences, and in the (future) case that there are more
+ * than these two levels for us to care about (e.g. HEADs in illumos-gate, ZoL
+ * and *BSD downstreams of OpenZFS) some interim version strings may be defined
+ * and if nothing is defined or value is unknown, then assume "legacy" mode.
+ */
+    private List<Entry<String, String>> libzfs_features;
+
+    String libzfs4j_features_get(String key) {
+        if (libzfs_features.size() > 0) {
+            for (Entry<String, String> entr : libzfs_features) {
+                if ( entr.getKey().equals(key) ) {
+                    return entr.getValue();
+                }
+            }
+        }
+        return "";
+    }
+
+    private String libzfs4j_get_extsetting(String libzfs4j_envvar_name) {
+        /* Ensure the caller has a string set via envvars or Java CLI.
+         * If the value is not set or is set to empty string, try the
+         * next option. If it ends up not set, return a "<NULL>". */
+        String libzfs4j_envvar_value;
+        libzfs4j_envvar_value = System.getenv(libzfs4j_envvar_name);
+        if (libzfs4j_envvar_value == null || libzfs4j_envvar_value.equals("")) {
+            libzfs4j_envvar_value = System.getProperty(libzfs4j_envvar_name);
+            if (libzfs4j_envvar_value == null || libzfs4j_envvar_value.equals("")) {
+                libzfs4j_envvar_value = "<NULL>";
```

Comment:
It requires more checks later on, e.g. `returnedvalue.equals("something")` bombed out when envvars were not set and the value ended up `null`.